File: Week3/Code/Vectorize1.R
#################
## Vectorize 1 ##
#################

## Author: Hannah O'Sullivan h.osullivan18@imperial.ac.uk
## Script: Vectorize1.R
## Desc: An exercise in vectorization
## Date: October 2018

#clear environment
rm(list = ls())

M <- matrix(runif(1000000),1000,1000)

SumAllElements <- function(M){
  Dimensions <- dim(M)
  Tot <- 0
  for (i in 1:Dimensions[1]){
    for (j in 1:Dimensions[2]){
      Tot <- Tot + M[i,j]
    }
  }
  return (Tot)
}

# Time this function
# Start the clock!
start_time <- as.numeric(proc.time()[3])

# Run function... tick tock
total <- (SumAllElements(M)) 

# Stop the clock!
end_time <- as.numeric(proc.time()[3])

# Get the elapsed time
elapsed_time <- end_time - start_time 

# Print the elapsed time. 
sprintf("The sum of all elements takes %ss to run.", round(elapsed_time, digits = 3))

File: Week3/Code/Vectorize1.py
#!/usr/bin/env python3
def main(argv=[]):

    """

    A python verison of Vectorize1.R

    """
    __author__ = "Hannah O'Sullivan (h.osullivan18@imperial.ac.uk)"
    __appname__ = "Vectorize 1"
    __version__ = "0.0.1"
    __license__ = "I do not have one"

    # Import modules
    import numpy as np
    import time

    # Define function
    def SumAllElements(M):
        """ A function to sum all the
        elements in a matrix """
        Dimensions = np.shape(M) # Get matrix dimensions
        Tot = 0 # Initialise
        for i in range(0, Dimensions[0]): # From the first...
            for j in range(0, Dimensions[-1]): # to last element
                Tot = Tot + M[i, j] # Sum these elements and add to the total
        return Tot

    # Draw random samples from a uniform distribution
    # With the dimmensions 1000x1000
    M = np.random.rand(1000, 1000)

    # Time the function:
    start = time.time() # Start the clock
    SumAllElements(M) # Run function
    # Stop the clock and print result.
    print("The sum of all elements takes %fs to run." % (time.time() - start))


if __name__== "__main__":
    import sys
    main(sys.argv)
